Infra server NP shim + config/routing API adoption

jasonrhodes opened the pull request:
Sets up a working NP shim for the server side of the infra plugin (logs and infra UI), implementing the NP routing and config management APIs as well.

Closes #44917

This PR implements the server side NP shim for the infra plugin, while also refactoring some of the abstraction code. It also migrates to the NP config and router.

jasonrhodes committed:
Basic cleanup before refactoring for shim work

jasonrhodes committed:
Basic cleanup before refactoring for shim work

jasonrhodes committed:
shim WIP

jasonrhodes committed:
shim WIP

jasonrhodes commented on x-pack/legacy/plugins/infra/server/new_platform_plugin.ts:
```diff
@@ -0,0 +1,52 @@
+        },
+      };
+    });
+  }
```
I'm not familiar with observables at all ... I guess we should be letting React components "subscribe" to the config so they get updates? I wonder if we can make that more react-friendly with a hook and hide the whole observable aspect in some cases...

jasonrhodes replied:
The trick will be how to do this kind of default value management, maybe need to wrap it in my own observable but I'm not sure.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2253/)

spalger commented:
Please merge master

jasonrhodes commented on x-pack/legacy/plugins/infra/index.ts:
```diff
@@ -71,7 +74,27 @@ export function infra(kibana: any) {
+          create: getConfig$,
+          createIfExists: getConfig$,
+        },
+      } as PluginInitializerContext;
```
This is a small shim for the config service observable layer. Definitely need some checks to see if this is correctly implemented and then maybe it could be moved somewhere central for others to use during migration.

jasonrhodes committed:
Removes the configuration adapter

jasonrhodes committed:
Removes the configuration adapter

rudolf reviewed (commented):
Left some comments, but this is definitely going in the right direction.

jasonrhodes committed:
WIP more stuff

jasonrhodes committed:
WIP more stuff

rudolf commented on x-pack/legacy/plugins/infra/index.ts:
```diff
@@ -71,7 +74,30 @@ export function infra(kibana: any) {
+      const kbnServer = (server as unknown) as KbnServer;
+      const { core } = kbnServer.newPlatform.setup;
+
+      const getConfig$ = <T>() =>
```
If anyone else looks to this as a reference, rather use the following:

```ts
const getConfig$ = new BehaviorSubject(
   legacyServer.config().get('xpack.infra')
).asObservable();
```

This avoids a bug in the previous suggestion that results in no config values if you subscribe to `getConfig$` before `server.config().get('xpack.infra')` resolves.

jasonrhodes committed:
WIP refactoring of shimming work

jasonrhodes committed:
WIP refactoring of shimming work

skh commented on x-pack/legacy/plugins/infra/server/new_platform_plugin.ts:
```diff
@@ -0,0 +1,102 @@
+        enabled: configValue.enabled,
+        query: {
+          ...DEFAULT_CONFIG.query,
+          ...configValue.query,
```
I found that depending on your schema, and whether config options are optional, you might end up with `undefined` values here.

jasonrhodes committed:
WIP continues

jasonrhodes committed:
WIP continues

jasonrhodes committed:
Logging UI now runs on top of new platform shim

jasonrhodes committed:
Logging UI now runs on top of new platform shim

jasonrhodes committed:
WIP continues

jasonrhodes committed:
WIP continues

jasonrhodes committed:
Removes unused imports and variables

jasonrhodes committed:
Removes unused imports and variables

jasonrhodes committed:
Basic infra NP server shim in place

jasonrhodes committed:
Basic infra NP server shim in place

jasonrhodes committed:
Reimplemented graphql http error handling for infra NP server shim

jasonrhodes committed:
Reimplemented graphql http error handling for infra NP server shim

jasonrhodes committed:
Adds new platform infra plugin to handle NP config for legacy server shim

jasonrhodes committed:
Adds new platform infra plugin to handle NP config for legacy server shim

jasonrhodes committed:
Adds comment about duplicating full config for NP config

Kerry350 committed:
Use New Platform features plugin to registerFeature()

jasonrhodes committed:
Fixing merge conflicts from collaboration

jasonrhodes committed:
Re-arranging and relying on request context as uch as possible

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8254/)
* Commit: 9982b9f97744d55718241a618da0c72deccfc78c

jasonrhodes committed:
Refactors KibanaRequest for RequestHandlerContext

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8420/)
* Commit: 4b7a64937477d4edd05323c9c1d79b205fce4041

jasonrhodes committed:
fixes types for callWithRequest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8445/)
* Commit: 645fa97a6d75ae45bcc84172330fb57f46e65404

jasonrhodes committed:
Moves callWithRequest method override types directly into class to get them working, need to fix this when we understand it better

jasonrhodes committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8464/)
* Commit: 07ad2809e467646489f37c1b081edbbec3d7c01a

jasonrhodes committed:
Fixes callWithRequest framework types

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8565/)
* Commit: 2cd167ac14dc7b6093342ac36541f4871504e87e

jasonrhodes committed:
Removes a few NP_TODO comments

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8622/)
* Commit: f530de16d4b35bafed1c71966488e48c3f77bc4a

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Fix broken imports

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/9050/)
* Commit: 1f0e119559967500dcffc9f7175c447ed7f9bea8

Kerry350 committed:
Ensure GraphQL resolvers are actually passed requestContext and not the raw request, and switch to the savedObjects client via requestContext

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/9142/)
* Commit: d9e616226f671e7062257dc9edb0f97991566cd6

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Remove the legacy traces of the savedObjects plugin

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/9418/)
* Commit: de9610e5e657e44e1de7a3590ff1214022a6b883

Kerry350 committed:
Fixes TSVB access with NP raw requests and requestContext

Kerry350 committed:
Remove unused getUiSettingsService (moved to requestContext)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/9954/)
* Commit: dbcc3aa2ec6eb24393603646cd3c5348daaaedf9

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Migrate to new Spaces plugin

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10025/)
* Commit: dce7b93a3eef54a6ce6e0ae9cda4af6d5faf1472

Kerry350 committed:
Fix calculateMetricInterval after merged changes

Kerry350 committed:
Reinstate and migrate the infrastructure metadata route

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10169/)
* Commit: c18b71dc65d82f1a6d2760ebe11ccd02d62afc75

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Fix various type check errors

Kerry350 committed:
Amend InfraSources lib unit tests

Mock the savedObjects client differently

Kerry350 committed:
Amend MetricsExplorer API response

Renaming of variable inadvertently broke the response

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10358/)
* Commit: b9565ed891e2d0a518dba60241460ea156223fd6

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10372/)
* Commit: fa8ddba718d3f35898eacad5052bdf6a6e233bee

elasticmachine commented:
Pinging @elastic/logs-metrics-ui (Team:logs-metrics-ui)

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Remove GraphQLI references from feature controls tests

Kerry350 committed:
Remove other GraphiQL references

Kerry350 committed:
Fix security / access issue

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10726/)
* Commit: 5d76435245380afca2ec2eb23dddfbcac5ba3729

Kerry350 committed:
Add a framework level registerRoute method which always adds access tags by default

Kerry350 committed:
*Temp* disable test

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10843/)
* Commit: 1805b06aef0911c6da611a8ae13c7e3da6774c13

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/10923/)
* Commit: 476734e6a443a9a8d0f348743cf80eabc7b29df4

Kerry350 commented:
I thought it might be worth trying to add a summary of the changes in this PR, as they are plentiful, and corresponding commits where things are relatively contained. This isn't necessarily exhaustive, but should help reviewers. There's also a [followup branch](https://github.com/Kerry350/kibana/tree/np-migration-followups) in progress that contains further enhancements that **aren't** considered blockers to this getting merged.

There is currently a singular test assertion disabled because I can't figure out what's happening there. But apart from that - AFAICT - everything is functional in the UI communicating with our server side code, tests are passing, and the build is green.

Non-exhaustive changes:

- Shims the server side of the Infra plugin
- Migrates to the NP router (and consequently switches as much code as possible to the use of `requestContext` over `request`)
- Migrates to the NP config
- Migrates to the NP features plugin (42cf3e5)
- Migrates to the NP spaces plugin (dce7b93)
- Migrates to the NP saved objects client (d9e6162 and 6319c1d) 
- Ensures all routes have access tags (5ca1e3b). On `master` we currently only have these assigned to the GraphQL route. 
- Removes GraphiQL
- Removes and unwinds some of our adapter abstractions

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'master' into logs-ui-server-np-shim

Kerry350 commented:
⚠️ Note to say this branch is now broken as the `getApmIndices()` method we're using from APM has now been altered to use the parameter `context: APMRequestHandlerContext`, which we're incompatible with. We should wait until this is officially exposed via `setup` in a public contract (i.e. `server.newPlatform.setup.plugins.apm.getApmIndices`.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11851/)
* Commit: f8723b42d43399d1548b678867f38e1813368f6c

Kerry350 committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Migrate the log rate validation endpoint to the new platform

Fully migrates the [Logs UI] log rate setup index validation #50008 PR to New Platform routing etc

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12163/)
* Commit: 32588c95d573efe09212e55b1d40df1b9ca9686f

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12386/)
* Commit: ec8bc6dee88f4d919dee9d31bfa1f939adb354d8

jasonrhodes commented:
Note: our APM dependency will be unbroken once this issue is tackled by the APM team: https://github.com/elastic/kibana/issues/49620#issuecomment-555061647

We may want to look and see if we can expose a type from their side in the meantime, so we can get this merged without being broken, assuming it will be properly fixed when that issue is done.

jasonrhodes committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 commented:
@jasonrhodes Unless I've misunderstood I don't think we can merge without the proper fix. The issue isn't so much the type, we could work around that, but what the expectation of that `APMRequestHandlerContext` argument is. It's an intersection of the standard `RequestHandlerContext` (which we could call the function with), and then augmented with extra properties which the context is extended with on the APM side. One of these is `config` which the function goes on to use here: `const apmIndicesConfig = getApmIndicesConfig(context.config);`. There's also `params`, `logger` and `__LEGACY`. I don't think we should try and make our `requestContext` match what's expected here, but instead wait for the public contract (which I'd likely expect to accept 0 arguments).

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Amend types

jasonrhodes committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/12956/)
* Commit: e0c8c1d856d02613888e2bbcf52422946a14381d

jasonrhodes committed:
Example of how to expose APM get indices method in NP

Kerry350 committed:
Fix calls to TSVB bug caused by object mutation
This is a temp fix as the TSVB NP migration will supercede this

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13223/)
* Commit: 17c6990a778b738a9c432d1e4e6974de5bd83d7d

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13230/)
* Commit: 294ffef834e67658c69c63449d8e8dc4ed53b98a

jasonrhodes committed:
Converts getApmIndices function to accept saved object client, implements usage in infra

jasonrhodes committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13318/)
* Commit: 21a6ceeb28f934645fa06832f523a469095c212b

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

Kerry350 committed:
Fix APM setup_request tests

jasonrhodes committed:
Fixes some unused references for linting

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13429/)
* Commit: 85fae624b3fd066085d538cc6912530919303a00

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13449/)
* Commit: 5e0cd63c7f8bf0c3fa778357acf37f7314598dec

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Migrate all work from #50730 to NP

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13648/)
* Commit: 6373386013820cd78be73c7bb949fdeaa4682d4d

Kerry350 committed:
Remove duplicate declaration files for rison_node and add a single source of truth at x-pack/typings/rison_node.d.ts for x-pack uses

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13709/)
* Commit: 46591d0e6ad512205bb8040dfbad8799e1054d1e

jasonrhodes committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

Kerry350 committed:
Merge remote-tracking branch 'upstream/master' into logs-ui-server-np-shim

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13880/)
* Commit: caba63f963887aacde17faf48cdcb9e27b17846a

jasonrhodes committed:
Moved type file back into infra plugin to bypass strange break

jasonrhodes committed:
Merge branch 'logs-ui-server-np-shim' of github.com:jasonrhodes/kibana into logs-ui-server-np-shim

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13907/)
* Commit: 78150b42eaa523e56e784554dca5b416c3be7183

jasonrhodes committed:
Updates apm indices method signature per feedback from @elastic/apm-ui

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/13943/)
* Commit: 6f37fecde7c077ebf68bfd54219053027fbeaf01

Kerry350 reviewed (approved):
Yay, approving for Logs / Metrics 🎉

sqren commented on x-pack/legacy/plugins/apm/server/lib/settings/apm_indices/get_apm_indices.ts:
```diff
@@ -54,15 +54,25 @@ export function getApmIndicesConfig(config: APMConfig): ApmIndicesConfig {
-export async function getApmIndices(context: APMRequestHandlerContext) {
+// export async function getApmIndices(context: APMRequestHandlerContext) {
+//   return _getApmIndices(context.core, context.config);
+// }
```
leftover?

Kerry350 replied:
I'll clean this up on the followup branch we have 👍
